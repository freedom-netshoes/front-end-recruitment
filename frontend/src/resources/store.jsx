import thunk from 'redux-thunk'
import { persistStore, autoRehydrate } from 'redux-persist'
import { compose, applyMiddleware, createStore } from 'redux'
import createCompressor from 'redux-persist-transform-compress'
import reducers from '../reducers/mainReducer'

// const devTools = window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()

// Be sure to ONLY add this middleware in development!
const middleware = process.env.NODE_ENV !== 'production'
  ? [require('redux-immutable-state-invariant').default(), thunk]
  : [thunk]

const compressor = createCompressor()
const middlewares = [...middleware]

export default function getStore (preloadedState) {
  const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
  const store = createStore(
    reducers,
    preloadedState,
    composeEnhancers(
      applyMiddleware(...middlewares),
      autoRehydrate()
    )
  )
  persistStore(store, { transforms: [compressor] })
  return store
}
