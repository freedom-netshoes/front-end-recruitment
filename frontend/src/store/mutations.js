export default {
  'MODAL' (state, payload) {
    state.modal = payload
  },
  'ADD_PROD_CART' (state, payload) {
    state.cart.show = payload.show
    payload.product.size = payload.size
    payload.product.quantify = payload.quantify
    state.cart.total += payload.product.price * payload.product.quantify
    state.modal.show = false
    state.cart.products.push(payload.product)
  },
  'REMOVE_PROD_CART' (state, payload) {
    state.cart.total -= payload.product.price * payload.product.quantify
    state.cart.products.splice(payload.index, 1)
  },
  'SHOW_AND_HIDE_CART' (state, payload) {
    state.cart.show = payload
  },
  'SNACK_BAR' (state, payload) {
    state.snackBar.show = true
    state.snackBar.message = payload.message
    state.snackBar.status = payload.status
    setTimeout(() => {
      state.snackBar.show = false
    }, payload.timer)
  },
  'BUY' (state, payload) {
    state.cart = {
      show: false,
      products: [],
      total: 0.00
    }
  }
}
