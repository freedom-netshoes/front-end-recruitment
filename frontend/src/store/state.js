export default {
  snackBar: {
    show: false,
    message: '',
    status: ''
  },
  modal: {
    show: false,
    product: null
  },
  cart: {
    show: false,
    products: [],
    total: 0.00
  }
}
