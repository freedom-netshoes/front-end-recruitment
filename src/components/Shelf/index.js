import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { fetchProducts } from '../../services/products/actions';

import Product from './Product';
import Clearfix from '../Clearfix';

class Shelf extends Component {

  componentWillMount() {
    this.handleFetchProducts();
  }

  handleFetchProducts = () => {
    this.props.fetchProducts();
  }

  render() {
    const { products } = this.props;

    const p = products.map(p => {
      return (
        <Product
          product={p}
          key={p.id}
        />
      );
    });

    return (
      <React.Fragment>
        <div className="shelf-container">
          {p}
          <Clearfix />
        </div>
        <Clearfix />
      </React.Fragment>
    )
  }
}

Shelf.propTypes = {
  fetchProducts: PropTypes.func.isRequired,
  products: PropTypes.array.isRequired,
}

const mapStateToProps = state => ({
  products: state.products.products,
})

export default connect(mapStateToProps, { fetchProducts })(Shelf);
