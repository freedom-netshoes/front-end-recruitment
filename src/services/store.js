import { compose, createStore, applyMiddleware } from "redux";
import thunk from 'redux-thunk';
import rootReducer from './reducers';

const initialState = JSON.parse(window.localStorage.getItem('state')) || {};
const middleware = [thunk];

const store = createStore(
  rootReducer,
  initialState,
  compose(
    applyMiddleware(...middleware),
    /* window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__() */
  )
);

store.subscribe(() => {
  const state = store.getState();
  const persist = {
    cart: state.cart,
    total: state.total,
  }

  window.localStorage.setItem('state', JSON.stringify(persist));
});

export default store;
