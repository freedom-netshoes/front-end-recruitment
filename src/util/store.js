export const itemInCart = function (cart, item) {
  let alreadyAdded = false
  for (let i = 0; i < cart.length; i++) {
    if (cart[i].sku === item.sku) {
      alreadyAdded = true
      break
    }
  }
  return alreadyAdded
}
