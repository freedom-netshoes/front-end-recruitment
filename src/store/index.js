'use strict'

import _axios from '../axios'
import { itemIsInCart } from '../util/store'

export const ADD_TO_CART = 'cart/ADD_TO_CART'
export const REMOVE_FROM_CART = 'cart/REMOVE_FROM_CART'

const state = {
  cart: [],
  products: [],
  error: null
}

const fetchProducts = function () {
  _axios.get('/products').then(resp => {
    state.products = resp.data
  })
}
fetchProducts()

const mutations = {
  [ADD_TO_CART] (state, item) {
    if (!itemIsInCart(state.cart, item)) {
      state.cart.push(item)
    }
  },
  [REMOVE_FROM_CART] (state, item) {
    state.cart.splice(state.cart.indexOf(item), 1)
  }
}

const getters = {
  cart: state => state.cart,
  products: state => state.products,
  error: state => state.error,
  productInCart: state => item => itemIsInCart(state.cart, item)
}

export default {
  state: state,
  getters: getters,
  mutations: mutations
}
