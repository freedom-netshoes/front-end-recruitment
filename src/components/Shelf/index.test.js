import Shelf from './';

import store from '../../services/store';

describe('Shelf', () => {
  
  it('should mount correctly', () => {
    const wrapper = mount(
      <Provider store={store}>
      <Shelf />
      </Provider>
    );
      
    expect(wrapper).toMatchSnapshot();
  });
});