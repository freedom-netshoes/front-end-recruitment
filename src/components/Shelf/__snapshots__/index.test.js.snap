// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shelf should mount correctly 1`] = `
Array [
  <div
    className="spinner lds-ring"
  >
    <div />
    <div />
    <div />
    <div />
  </div>,
  <div
    className="filters"
  >
    <h4
      className="title"
    >
      Tamanhos:
    </h4>
    <div
      className="filters-available-size"
    >
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
          value="40"
        />
        <span
          className="checkmark"
        >
          40
        </span>
      </label>
    </div>
    <div
      className="filters-available-size"
    >
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
          value="41"
        />
        <span
          className="checkmark"
        >
          41
        </span>
      </label>
    </div>
    <div
      className="filters-available-size"
    >
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
          value="43"
        />
        <span
          className="checkmark"
        >
          43
        </span>
      </label>
    </div>
    <div
      className="filters-available-size"
    >
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
          value="S"
        />
        <span
          className="checkmark"
        >
          S
        </span>
      </label>
    </div>
    <div
      className="filters-available-size"
    >
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
          value="G"
        />
        <span
          className="checkmark"
        >
          G
        </span>
      </label>
    </div>
    <div
      className="filters-available-size"
    >
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
          value="GG"
        />
        <span
          className="checkmark"
        >
          GG
        </span>
      </label>
    </div>
    <div
      className="filters-available-size"
    >
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
          value="GGG"
        />
        <span
          className="checkmark"
        >
          GGG
        </span>
      </label>
    </div>
  </div>,
  <div
    className="shelf-container"
  >
    <div
      className="shelf-container-header"
    >
      <small
        className="products-found"
      >
        <span>
          0
           Produto(s) encontrado.
        </span>
      </small>
      <div
        className="sort"
      >
        Ordenar por 
        <select
          onChange={[Function]}
        >
          <option
            key=""
            value=""
          >
            Selecionar
          </option>
          <option
            key="lowestprice"
            value="lowestprice"
          >
            Menor Preço
          </option>
          <option
            key="highestprice"
            value="highestprice"
          >
            Maior Preço
          </option>
        </select>
      </div>
      <div
        className="clearfix"
      />
    </div>
    <div
      className="clearfix"
    />
  </div>,
  <div
    className="clearfix"
  />,
]
`;
